firmware/application/ui/ui_textentry.cpp

```diff
@@ -45,76 +45,191 @@ void text_prompt(NavigationView& nav, std::string& str, const size_t max_length,
 	}*/
 }
 
-void TextEntryView::update_text() {
-	if (cursor_pos < 30)
-		text_input.set(_str + std::string(_max_length - _str.length(), ' '));
+/* TextField ***********************************************************/
+
+TextField::TextField(
+	std::string& str,
+	size_t max_length,
+	Point position,
+	uint32_t length
+) : Widget{ { position, { 8 * static_cast<int>(length), 16 } } },
+	text_{ str },
+	max_length_{ std::max<size_t>(max_length, 1) },
+	char_count_{ std::max<uint32_t>(length, 1) },
```

Comment:
Should test for the upper bound too, but not critical.

Reply:
Can you add the missing check please ?

Reply:
Or is it unneeded ?

Reply:
It’s not actually needed and it’s a hypothetical check anyway. All it would do would try drawing off the end of the screen and that’s only if someone were to use this Widget directly.